### ISSUE TITLE
Underscore key support for environment variables

### ISSUE THREAD

**Issue by ProTip** [kelseyhightower/confd#737]
Unfortunately the `env` back-end makes asthetic variable naming a bit.. tricky. This is due to all `_` being replaced with `/`, as well as the key upper being compared to the envar raw(no room for casing).

 #567 attempts to solve this by through the key naming mechanism, however a "gotcha" was pointed out as well when using key prefixes. I would like to propose an alternative approach:

**`__Dunder__`**
We can replace only single `_` in the envars with `/`, and shorten dunder `__` to underscore `_` for comparisons.

`APP_SERVER_BIND__PORT` => `app/server/bind_port`

I believe this also has the potential for a negligable impact on backwards compatibility as previously a dunder would have been converted to a das.. `//` 😄

The primary motivation for me on this is to allow for a more "conventional" looking naming scheme for environment variables while also playing well with alternative backends. I would be happy to submit a PR if this approach sounds good.

**Comment by hubo1016** [comment]
It is not only underscore `_` that is causing the problem, many other characters like `-` and `:` are not supported in environment variables. We can only say that this is a very limited backend. Currently I'm using YAML file instead - put a YAML file completely in the environment variable (e.g. when creating a container), save it to a YAML file in a starting script, and use confd file backend.

**Comment by ProTip** [comment]
Yes, I can agree it is certainly more limited in keyspace than most(all?) of the other backends. However, I think more can be done than simply pointing this out 😄 

In my case I'm distributing docker images for others to use outright or build on top of. While cramming a document into an environment variable may work for my own deployment purposes, it's not such a good look for a distributed image.

What I would like is a way for people to start out with a Docker Hub image, try it out using conventional configuration means(environment variables and the odd volume where absolutely necessary), and then have a clear path forward for more advanced configuration as their needs and deployment infrastructure mature.

As it is, environment variables are sort of the "defacto", familiar configuration API for docker containers and basic compose setups. While `APP_JAAS_MODULE_JETTY__CACHING__LDAP` is _slightly_ unconventional, it is much nicer than `APP_JAAS_MODULE_JETTYCACHINGLDAP`.

**Comment by hubo1016** [comment]
Different people may have different feelings for this... Some may think that two underscores are more "separated" than one underscore, so they may feel strange that why `APP_MYAPP_CACHE__BACKUP` corresponds to `/app/myapp/cache_backup` but not `app_myapp_cache/backup`. Env backends should be very straight forward, so I'm afraid that this design - although works quite well - may be confusing for users.
